lookout/style/typos/analyzer.py

```diff
@@ -273,20 +324,36 @@ def reconstruct_identifier(tokenizer: TokenParser, pred_tokens: List[str], ident
         return "".join(res)
 
     @classmethod
+    @with_uasts_and_contents(unicode=True)
     def train(cls, ptr: ReferencePointer, config: Mapping[str, Any], data_service: DataService,
-              **data) -> AnalyzerModel:
+              files: Iterator[File], **data) -> AnalyzerModel:
         """
         Generate a new model on top of the specified source code.
 
         :param ptr: Git repository state pointer.
         :param config: Configuration of the training of unspecified structure.
         :param data_service: The channel to the data service in Lookout server to query for \
                              UASTs, file contents, etc.
+        :param files: iterator of File records from the data service.
         :param data: Extra data passed into the method. Used by the decorators to simplify \
                      the data retrieval.
         :return: Instance of `AnalyzerModel` (`model_type`, to be precise).
         """
-        return DummyAnalyzerModel()
+        _log = logging.getLogger(cls.__name__)
+        train_config = cls._load_config(config)["train"]
+        _log.info("train %s %s %s %s", __version__, ptr.url, ptr.commit,
+                  pformat(train_config, width=4096, compact=True))
+        model = IdTyposModel()
+        for _, files in files_by_language(files).items():
+            for file in filter_files(
+                    files=files, line_length_limit=train_config["line_length_limit"],
+                    overall_size_limit=train_config["overall_size_limit"], log=_log):
+                model.update({
+                    node.token for node in uast2sequence(file.uast)
+                    if bblfsh.role_id("IDENTIFIER") in node.roles
```

Comment:
`bblfsh.role_id("IDENTIFIER")` should be at top function level - it takes time for every call. The same about `IMPORT`